stats/src/main/java/com/facebook/airlift/stats/QuantileDigest.java

```diff
@@ -1315,4 +1337,172 @@ public String toString()
+        private final double[] counts; // counts of nodes
+        private final byte[] levels; // levels in tree of nodes
+        private final long[] values; // values of nodes
+        private final double weightedCount; // total count of tree nodes
+        private final long min;
+        private final long max;
+        private final BitSet seen;
+
+        private int currentNode = -1; // current node in tree
+        private int stackPosition = -1; // current head position in stack array
+        private double cumulativeCount; // cumulative count of nodes in iteration
+        private double previousCumulativeCount; // previous value of cumulative count (0 if first element)
+        private boolean advanced; // iterator position state
+
+        public QuantileDigestIterator(int root, int[] lefts, int[] rights, double[] counts, byte[] levels, long[] values, double weightedCount, long min, long max)
+        {
+            this.weightedCount = weightedCount;
+            this.lefts = lefts;
+            this.rights = rights;
+            this.counts = counts;
+            this.levels = levels;
+            this.values = values;
+            this.min = min;
+            this.max = max;
+            seen = new BitSet(values.length);
+            stack = new int[(Integer.highestOneBit(values.length - 1) << 1) + 2]; // log2 ceiling
+            push(root);
+        }
+
+        public long upperBoundValue()
+        {
+            validateCurrentNodePosition();
+            return Math.min(upperBound(currentNode, levels, values), max);
+        }
+
+        public long lowerBoundValue()
+        {
+            validateCurrentNodePosition();
+            return Math.max(lowerBound(currentNode, levels, values), min);
+        }
+
+        public double count()
+        {
+            validateCurrentNodePosition();
+            return counts[currentNode];
+        }
+
+        public double cumulativeCount()
+        {
+            validateCurrentNodePosition();
+            return cumulativeCount;
+        }
+
+        public double lowerBoundQuantile()
+        {
+            validateCurrentNodePosition();
+            return previousCumulativeCount / weightedCount;
+        }
+
+        public double upperBoundQuantile()
+        {
+            validateCurrentNodePosition();
+            return cumulativeCount / weightedCount;
+        }
+
+        public boolean hasNext()
+        {
+            if (advanced) {
+                return stackIsNotEmpty();
+            }
+            moveToNextNode();
+            advanced = true;
+            return stackIsNotEmpty(); // if we advanced to the next value, we will always be at a valid position in the stack
+        }
+
+        public void advance()
+        {
+            if (!hasNext()) {
+                throw new NoSuchElementException();
+            }
+            advanced = false;
+        }
+
+        private void moveToNextNode()
+        {
+            while (stackIsNotEmpty()) {
```

Comment:
Add some brief comments about what you are doing here? :) (e.g. trarverse to the next node in post-order)